src/transformers/models/mgp_str/modeling_mgp_str.py

```diff
@@ -0,0 +1,363 @@
+
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, config.embed_dim))
+        self.patch_embed = MGPSTRPatchEmbeddings(config.img_size, config.patch_size, config.in_chans, config.embed_dim)
+        num_patches = self.patch_embed.num_patches
+        self.pos_embed = nn.Parameter(torch.zeros(1, num_patches + self.num_tokens, config.embed_dim))
+        self.pos_drop = nn.Dropout(p=config.drop_rate)
+
+        self.encoder = MGPSTREncoder(
+            config.embed_dim,
+            config.depth,
+            config.num_heads,
+            config.mlp_ratio,
+            config.qkv_bias,
+            config.drop_rate,
+            config.attn_drop_rate,
+            config.drop_path_rate,
+            config.norm_layer,
+            config.act_layer,
+        )
+
+        self.char_a3_module = MGPSTRA3Module(config.embed_dim, config.max_token_length)
+        self.bpe_a3_module = MGPSTRA3Module(config.embed_dim, config.max_token_length)
+        self.wp_a3_module = MGPSTRA3Module(config.embed_dim, config.max_token_length)
+
+        self.char_head = nn.Linear(config.embed_dim, config.char_num_classes)
+        self.bpe_head = nn.Linear(config.embed_dim, config.bpe_num_classes)
+        self.wp_head = nn.Linear(config.embed_dim, config.wp_num_classes)
+
+    def forward(self, pixel_values):
+        r"""
+        Args:
+            pixel_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+                Pixel values. Pixel values can be obtained using [`AutoImageProcessor`]. See
+                [`ViTImageProcessor.__call__`] for details.
+        Returns:
+            out(`List[torch.FloatTensor]`): The list of logits output of char, bpe, wp.
+            attens(`List[torch.FloatTensor]`): The list of attention output of char, bpe, wp.
+
+        Example:
+
+        ```python
+        >>> from transformers import (
+        ...     MGPSTRModel,
+        ...     MGPSTRProcessor,
+        ... )
+        >>> import requests
+        >>> from PIL import Image
+
+        >>> # load image from the IIIT-5k dataset
+        >>> url = "https://i.postimg.cc/ZKwLg2Gw/367-14.png"
+        >>> image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+
+        >>> processor = MGPSTRProcessor.from_pretrained("alibaba-damo/mgp-str-base")
+        >>> pixel_values = processor(images=image, return_tensors="pt").pixel_values
+
+        >>> model = MGPSTRModel.from_pretrained("alibaba-damo/mgp-str-base")
+
+        >>> # inference
+        >>> generated_ids, attens = model(pixel_values)
+        >>> outs = processor.batch_decode(generated_ids)
+        >>> outs["generated_text"]
+        '['ticket']'
+        ```"""
+
+        if pixel_values is None:
+            raise ValueError("You have to specify pixel_values")
+
+        B = pixel_values.shape[0]
+        x = self.patch_embed(pixel_values)
+
+        cls_tokens = self.cls_token.expand(B, -1, -1)  # stole cls_tokens impl from Phil Wang, thanks
```

Comment:
Rather than say you 'stole' it would be good to link to the code implementation where it's from.